### DISCUSSION TITLE
PTB high entry threshold by number of syntax sugar and features.

### DISCUSSION THREAD

**Discussion by david-shiko** [python-telegram-bot/python-telegram-bot#2801] [Q&A]
During a more deepth investigate of  PTB, I noticed that it's growing and flowering. 
But at the same time it's becoming to be a little overbloated and sometimes (IMO) has a redundant syntax sugar or features. 
May seem that such sugar and numbre of features is helps to PTB users, but this also increases the entry threshold. 
I noticed this moment when started to investigate benefits of `extbot` over a `bot`. 
And [this](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Arbitrary-callback_data) papper particulary.
As I understand from this, PTB user may bind a function to Inline button. 
In reality instead of function object will be sent his uuid (so somwhere in memory there a mapping between callback functions and uuid).
Is it not easier to just use a regular regex and run this function manually?
Even if papper above is a bad examples,  a question is still actually.
I pretty understand a wish of PTB developers to grow, but growing sake of growing is always bad. Some things may preserve a perfect just because they are not changing :)

**Comment by Poolitzer** [comment]
> Is it not easier to just use a regular regex and run this function manually?

Arguably not? You can not attach a function anyway, but any object, which simplifies user code a lot, though it might have a bit of a complex backend. But being able to do:

```python
user = UserObject(name="General Kenobi", colour="blue", catchphrase="Hello there")
InlineButton("Press this, User X", callback_data=user)
```

and then in the next function

```python
send_message(update.chat.id, f"You are a bold one, {update.callback_data.name}, with your {update.callback_data.colour} lightsaber and your great catchphrase {update.callback_data.catchphrase}")
```

feels really good in code, at least to me.

Maybe you have other examples you mean?

If you speak from a user perspective and not a developer one, there is [python-telegram-bot-raw](https://github.com/python-telegram-bot/python-telegram-bot/blob/master/README_RAW.rst), which is only the API into nice python objects.


We do try to lower the barrier of entry for users, we provide copy paste examples in a folder, lots of (hopefully) easy to understand wiki pages with introductions, and a lot of interlinking in our doc at Read the Docs. Though a documentation always can be and should be improved, so I am very happy if you can point me to things you found hard to read/learn about (see also #2633)

**Reply by david-shiko** [reply]
Thanks for a userful link

**Comment by Bibo-Joshi** [comment]
Hi, thanks for reaching out and giving feedback - after all, that's the only way we can improve PTB :)

Adding to @Poolitzer, I'd like to explain that new features almost exclusively come to be because users suggest them or because the developer team notices that a new feature could be helpful for users. E.g. in the particular case of arbitrary callback data, the question on how to pass larger datasets through inline buttons and the suggested answer to store the data using a unique identifier came up multiple times in our user group on Telegram.
But ofc I agree that arbitrary callback data is one of the most advanced features of PTB and I also agree that adding features for the sake of having features is not healthy. In fact, we have a dedicated repository [ptbcontrib](../../ptbcontrib) for features that the dev team deems to specific to maintain within the main library. And if we realize that a feature takes too much work to maintain, we have reconsider. E.g. for v13, we decided to no longer implement `JobQueue` manually but to rely on a 3rd party backend for the scheduling instead, which eased maintenance a lot.

You said that the number of feature increases the entry threshold and here I'd like to repeat Poolitzers invitation to tell us how we can improve documentation & other resoucres. Almost all features of `telegram.ext` are opt-in only, specifically the features of `ExtBot` (namly arbitrary callback data and `Defaults`), which is why I'm wondering how they increase the entry threshold and how we can avoid that. I'd be glad for feedback on that :)
